tests/cxx/tests/DREAM/MeanExcitationEnergy.cpp

```diff
@@ -0,0 +1,198 @@
+#include "DREAM/Settings/OptionConstants.hpp"
+#include "DREAM/Equations/CoulombLogarithm.hpp"
+#include "DREAM/Equations/SlowingDownFrequency.hpp"
+#include "DREAM/Equations/PitchScatterFrequency.hpp"
+#include "FVM/UnknownQuantityHandler.hpp"
+#include "FVM/Grid/Grid.hpp"
+#include <iostream>
+
+using namespace DREAMTESTS::_DREAM;
+using namespace std;
+
+
+/**
+ * Run this test.
+ */
+bool MeanExcitationEnergy::Run(bool) {
+
+    bool success = true;
+    if (CompareMeanExcitationEnergyWithTabulated())
+        this->PrintOK("The mean excitation energy agrees with tabulated values.");
+    else {
+        success = false;
+        this->PrintError("The mean excitation energy test failed.");
+    }
+    return success;
+}
+
+
+/**
+ * Generate an unknown quantity handler.
+ */
+DREAM::FVM::UnknownQuantityHandler *MeanExcitationEnergy::GetUnknownHandler(DREAM::FVM::Grid *g, 
+    const len_t N_IONS, const len_t *Z_IONS, const real_t ION_DENSITY_REF, const real_t T_cold) {
+    DREAM::FVM::UnknownQuantityHandler *uqh = new DREAM::FVM::UnknownQuantityHandler();
+
+    len_t nZ0 = 0;
+    for (len_t i = 0; i < N_IONS; i++)
+        nZ0 += Z_IONS[i] + 1;
+
+    this->id_ions = uqh->InsertUnknown(DREAM::OptionConstants::UQTY_ION_SPECIES, "0", g, nZ0);
+    uqh->InsertUnknown(DREAM::OptionConstants::UQTY_N_COLD, "0", g);
+    uqh->InsertUnknown(DREAM::OptionConstants::UQTY_N_HOT, "0", g);
+    uqh->InsertUnknown(DREAM::OptionConstants::UQTY_N_TOT, "0", g);
+    uqh->InsertUnknown(DREAM::OptionConstants::UQTY_T_COLD, "0", g);
+    uqh->InsertUnknown(DREAM::OptionConstants::UQTY_F_HOT, "0", g);
+    uqh->InsertUnknown(DREAM::OptionConstants::UQTY_E_FIELD, "0", g);
+    
+
+    real_t ni;
+    // Set initial values
+    // we don't care about the density, just set it to the same value everywhere.
+    const len_t N = nZ0*g->GetNr();
+    real_t *nions = new real_t[N];
+    real_t ncold = 0;
+    real_t ntot = 0;
+    len_t ionOffset = 0, rOffset = 0;
+    for (len_t iIon = 0; iIon < N_IONS; iIon++) 
+        for (len_t Z0 = 0; Z0 <= Z_IONS[iIon]; Z0++, ionOffset++){ 
+            ni =  ION_DENSITY_REF;
+            ncold += Z0*ni;
+            ntot  += Z_IONS[iIon]*ni;
+            for (len_t ir = 0; ir < g->GetNr(); ir++, rOffset++)
+                nions[rOffset] = ni;
+        }
+        
+    
+    uqh->SetInitialValue(DREAM::OptionConstants::UQTY_ION_SPECIES, nions);
+
+    #define SETVAL(NAME, v) do { \
+            for (len_t i = 0; i < g->GetNr(); i++) \
+                temp[i] = (v); \
+            uqh->SetInitialValue((NAME), temp); \
+        } while (false)
+
+    // Set electron quantities
+    real_t *temp = new real_t[g->GetNr()];
+    SETVAL(DREAM::OptionConstants::UQTY_N_COLD, ncold);
+    SETVAL(DREAM::OptionConstants::UQTY_N_HOT,  ncold*1e-12);
+    SETVAL(DREAM::OptionConstants::UQTY_N_TOT,  ntot);
+    SETVAL(DREAM::OptionConstants::UQTY_T_COLD, T_cold);
+    SETVAL(DREAM::OptionConstants::UQTY_F_HOT, 0);
```

Comment:
do all of these actually need to be initialised? I thought you only needed n_cold, ion_species and T_cold for this problem?

Reply:
Fixed! Thanks!